Add text-to-speech (TTS) functionality for currently selected text (issue #5)

michaelweghorn opened the pull request:
This adds a basic text-to-speech (TTS) functionality for
the currently selected text, using QTextToSpeech.

Speech output can be triggered as follows:

* select text
* open context menu buy right-clicking on the selected text
* select the newly added  "Speak selected text" entry

In order to have QTextToSpeech use the proper language,
a locale generated from the text language is set.

Currently, only the locale is explicitly set for QTextToSpeech,
which means that the default voice for the given locale is used.
(This could be made configurable in a follow-up step to e.g.
allow the user to specify what voice to use for what language in
the application settings.)

The newly needed Qt5TextToSpeech module is added in
cmake/BTApplication.cmake accordingly.

Related: Issue #5

jaakristioja commented on cmake/BTApplication.cmake:
```diff
@@ -9,6 +9,7 @@ FIND_PACKAGE(Qt5Gui ${REQUIRED_QT_VERSION} REQUIRED)
 FIND_PACKAGE(Qt5Widgets ${REQUIRED_QT_VERSION} REQUIRED)
 FIND_PACKAGE(Qt5Xml ${REQUIRED_QT_VERSION} REQUIRED)
 FIND_PACKAGE(Qt5Test ${REQUIRED_QT_VERSION} REQUIRED)
+FIND_PACKAGE(Qt5TextToSpeech ${REQUIRED_QT_VERSION} REQUIRED)
```
Can this be made configurable using a compile-time flag to CMake?

michaelweghorn replied:
Done, there's now `-DBUILD_TEXT_TO_SPEECH=OFF` to disable the feature.

jaakristioja commented on src/frontend/display/btmodelviewreaddisplay.cpp:
```diff
@@ -207,6 +208,24 @@ void BtModelViewReadDisplay::print(TextPart const type,
+
+    // use language from module for TTS
+    const int selectedColumnIndex = m_parentWindow->getSelectedColumn();
+    assert(m_parentWindow->modules().size() > selectedColumnIndex);
```
Please use `BT_ASSERT` instead of regular `assert`.

michaelweghorn replied:
Done.

jaakristioja commented on src/frontend/display/btmodelviewreaddisplay.cpp:
```diff
@@ -207,6 +208,24 @@ void BtModelViewReadDisplay::print(TextPart const type,
 
+void BtModelViewReadDisplay::speakSelectedText() {
+    if (!m_textToSpeech)
+        m_textToSpeech.reset(new QTextToSpeech);
```
Please instead use `m_textToSpeech = std::make_unique<QTextToSpeech>();`

michaelweghorn replied:
Done

jaakristioja commented on src/frontend/display/btmodelviewreaddisplay.h:
```diff
@@ -207,4 +211,5 @@ class BtModelViewReadDisplay : public QWidget {
     BtQmlScrollView* m_widget;
     QString m_currentAnchorCache;
 
+    std::unique_ptr<QTextToSpeech> m_textToSpeech;
```
I think it might make more sense to create a single `QTextToSpeech` instance for the whole application, instead one for each `BtModelViewReadDisplay`. I suggest moving this to the main window class (`BibleTime`).

michaelweghorn replied:
Done.

jaakristioja commented on src/frontend/displaywindow/cdisplaywindow.cpp:
```diff
@@ -279,6 +279,9 @@ void CDisplayWindow::insertKeyboardActions( BtActionCollection* a ) {
     actn->setShortcut(CResMgr::displaywindows::general::findStrongs::accel);
     a->addAction(CResMgr::displaywindows::general::findStrongs::actionName, actn);
+
+    actn = new QAction(QIcon(), tr("Speak selected text"), a);
```
The `QIcon()` argument is not needed here.

michaelweghorn replied:
Indeed. I've dropped it.

jaakristioja commented on src/frontend/display/btmodelviewreaddisplay.cpp:
```diff
@@ -207,6 +208,24 @@ void BtModelViewReadDisplay::print(TextPart const type,
+
+    std::shared_ptr<const Language> lang = module->language();
+    const QLocale locale = QLocale(lang->abbrev());
+    m_textToSpeech->setLocale(locale);
```
These three (or more) lines can be converted into a single statement to avoid extending the lifetime of the `lang` and `locale` variables.

michaelweghorn replied:
Automatic locale-handling here was removed in the new version of the PR altogether, in favor of being able to explicitly configure the engine, locale, and voice.

jaakristioja commented on src/frontend/display/btmodelviewreaddisplay.cpp:
```diff
@@ -207,6 +208,24 @@ void BtModelViewReadDisplay::print(TextPart const type,
+    m_textToSpeech->setLocale(locale);
+
+    const QString currentText = qmlInterface()->getSelectedText();
+    m_textToSpeech->say(currentText);
```
This can also be a single statement, i.e. `m_textToSpeech->say(qmlInterface()->getSelectedText());`

jaakristioja reviewed (changes requested):
I tested this, but it did not output any audio. Only after I removed the logic calling `m_textToSpeech->setLocale()` it began to work.

This is a really nice start! 😃 But I think we need a bit more to introduce such a new feature into BibleTime. I think that at minimum the following should implemented:

- This feature should be made optional at compile-time, but enabled by default.
- The user should have the option to choose both the language and voice used for synthesizing the text.
- Corner cases like no languages/voices being available, or the user-selected language/voice not being available after BibleTime restart etc must be gracefully handled, possibly notifying and instructing the user when errors occur.
- I think there should only be one application-wide instance of `QTextToSpeech`, probably in `class BibleTime` (the main window).

Please feel free to ask, discuss and convince me otherwise. 😎

michaelweghorn commented:
Sorry for the extremely long delay before coming back to this! I've finally pushed an update that tries to address all the review comments.

> I tested this, but it did not output any audio. Only after I removed the logic calling `m_textToSpeech->setLocale()` it began to work.

I could not reproduce this on Debian testing. The new version allows explicitly setting a locale from those that the `QTextToSpeech` instance supports instead of trying to create a locale from the language of the current bible translation. Does that fix the issue for you?

> This is a really nice start! 😃 But I think we need a bit more to introduce such a new feature into BibleTime. I think that at minimum the following should implemented:
> 
>     * This feature should be made optional at compile-time, but enabled by default.

Disabling the feature is now possible using the `-DBUILD_TEXT_TO_SPEECH=OFF` CMake switch.

>     * The user should have the option to choose both the language and voice used for synthesizing the text.

I've added a new "Text-to-speech" section to the app settings that allows this, and also selecting a TTS engine, if multiple are available. Engines are provided by corresponding QTextToSpeech modules. On Debian, two of them can e.g. be installed via the packages `qt6-speech-flite-plugin` and  `qt6-speech-speechd-plugin` (for Qt 6) or `qtspeech5-flite-plugin` and `qtspeech5-speechd-plugin` (for Qt 5).
For now, the configured voice is used for all Bible translations. (This could e.g. be extended to allow overriding the default voice for specific languages, e.g. to configure an English voice by default, but set a different voice for German Bible translations, so there's no need to change the settings whenever switching between Bible translations of  different languages, but that's maybe more of a corner case and something I think can be seen as a feature that could still be implemented later, independent of this initial PR.)

>     * Corner cases like no languages/voices being available, or the user-selected language/voice not being available after BibleTime restart etc must be gracefully handled, possibly notifying and instructing the user when errors occur.

The current version of the PR saves the config and tries to restore it, but only does so if the configured engine/locale/voice is actually available, and falls back to defaults otherwise.

>     * I think there should only be one application-wide instance of `QTextToSpeech`, probably in `class BibleTime` (the main window).

That makes sense. I've moved it to the `BibleTime` class now.

michaelweghorn commented:
I have rebased this onto current git master now and fixed the merge conflicts.

jaakristioja commented on src/frontend/settingsdialogs/bttexttospeechsettings.cpp:
```diff
@@ -0,0 +1,152 @@
+
+    const QStringList availableEngines = QTextToSpeech::availableEngines();
+    if (availableEngines.isEmpty()) {
+        formLayout->addWidget(new QLabel(tr("No Qt text-to-speech plug-ins were found. Please install a plugin and restart the application."), this));
```
Please try to adhere to a 80-character line length limit.

michaelweghorn replied:
Done

jaakristioja reviewed (changes requested):
I have yet to run this, but overall the code looks very solid and thought through. In fact it seems to be of better quality that many other code in BibleTime. I'm very positive about this. 👍🏻 

Besides what I've voiced in other comments is that I think the code should make more use of `auto`, e.g. for lambda parameters, initializations etc. Spelling out an type name instead of using `auto` is in most cases only needed when there is uncertainty about the type of value initializing the variable and the value has to be coerced to a specific type.

The rest of the concerns stated below in this comment are not blockers, so please don't feel disheartened by these.

Personally I prefer not to expose private methods in class definitions if they don't actually need to be methods, and can be migrated to anonymous namespaces, converted to lambdas or just inlined. On the other hand there can of course be reasonable exceptions to this.

Your code style is nice and consistent, but slightly different from what we've tried to use in the rest of the code. I apologize that we don't currently have a code style guide. I can reformat it after merging, so you don't need to bother yourself with this unless you want to. Here are some examples:

- In class definitions we try to use `private: // methods:`,  `private: // fields` and similar.
- We try to use "east const" style, e.g. `QLocale const & locale`.
- Spaces round `&` and `*` in types, e.g. `int * i;`.
- `void f() {\n` instead of `void f()\n{\n` unless the parameter list spans multiple lines.
- If possible, then we usually omit the `{}` braces around the body of the code block, except for `else` and for `if` if followed by `else`.

michaelweghorn commented:
Thanks a lot for the very kind and constructive review!

I've pushed a new version that includes another rebase onto current master including conflict resolution and tries to address the review comments. I've marked most of them as resolved and added questions for 2 for clarification.

> Your code style is nice and consistent, but slightly different from what we've tried to use in the rest of the code. I apologize that we don't currently have a code style guide. I can reformat it after merging, so you don't need to bother yourself with this unless you want to. Here are some examples:
> [...]

I've tried to adjust the code accordingly, but most likely still missed something.

michaelweghorn commented on src/frontend/settingsdialogs/bttexttospeechsettings.cpp:
```diff
@@ -0,0 +1,152 @@
+            QTextToSpeech candidate = QTextToSpeech(engineName);
+            if (candidate.availableLocales() == m_tts->availableLocales()) {
+                candidate.setLocale(m_tts->locale());
+#if QT_VERSION >= QT_VERSION_CHECK(6, 0, 0)
```
Indeed. Adjusted now, and also dropped the code for Qt < 6.4, as that is now also no longer supported since

        commit 5fb6b5d7162eba53714a31d25f690e3efb47c9fb
        Author: Jaak Ristioja <jaak@ristioja.ee>
        Date:   Sat Mar 22 13:56:37 2025 +0200
    
            Drop support for Qt before version 6.7

michaelweghorn commented on pics/icons/audio.svg:
Would https://invent.kde.org/frameworks/oxygen-icons/-/blob/master/applets/22x22/speaker-symbolic.svg be appropriate? If so, what's the correct way to include it here? (Just copy it over + mention in the commit message where the icon comes from?)

jaakristioja replied:
I think it were better to use [this icon](https://invent.kde.org/frameworks/oxygen-icons/-/blob/master/scalable/apps/preferences-desktop-text-to-speech.svgz) which is specifically meant for preferences and fits the overall theme a little better with its color.

I just read through [this](https://techbase.kde.org/Projects/Oxygen/Licensing) and I think the BibleTime code does not fully meet the conditions. I think the other icons were added when there was some confusion about whether BibleTime is GPLv2 or GPLv2+. I should clear this up before we add another icon.

jaakristioja replied:
I don't think it makes sense for me to block this PR simply because of the current mess.

I guess public domain is okay, but [actual license on the website](https://publicicons.org/license) is a modified version of the [Unlicence](https://unlicense.org/) licence, and the additions are suspicious. The repository providing the actual files instead seems to be licensed under [CC0 1.0 Universal](https://github.com/davidmerfield/Public-Icons/blob/master/LICENSE). This also is suspicious. If you would provide your own it could be under the same GPL-2.0-only as the rest of the BibleTime code.

michaelweghorn commented on src/frontend/settingsdialogs/bttexttospeechsettings.cpp:
```diff
@@ -0,0 +1,152 @@
+#else
+    // QTextToSpeech::engine() only available in Qt >= 6.4, identify
+    // current engine by available locales and voices
+    if (availableEngines.size() > 1) {
```
The idea here was to preselect the correct engine if multiple engines are available. I don't think anything would have been necessary if there's no or just a single engine, as no engine other than the (single) default engine could have been selected in that case. In any case, the code is gone now, since it was only needed for older and now no longer supported Qt versions.

michaelweghorn commented on src/frontend/settingsdialogs/bttexttospeechsettings.cpp:
```diff
@@ -0,0 +1,152 @@
+        if (locale == m_tts->locale()) {
+            m_ttsLocaleComboBox->setCurrentIndex(m_ttsLocaleComboBox->count() - 1);
+        }
+    }
```
I hadn't, but adjusted the code now. Does that address this comment?

michaelweghorn commented on src/frontend/settingsdialogs/bttexttospeechsettings.cpp:
```diff
@@ -0,0 +1,152 @@
+    QVector<QLocale> locales = m_tts->availableLocales();
+    std::sort(locales.begin(), locales.end(),
+              [](const QLocale& locale1, const QLocale& locale2) {
+                return formatLocale(locale1) < formatLocale(locale2);
```
Good point. I've adjusted it to use `QString::localeAwareCompare` now.

michaelweghorn commented on src/frontend/settingsdialogs/bttexttospeechsettings.cpp:
```diff
@@ -0,0 +1,152 @@
+
+QString BtTextToSpeechSettingsPage::formatLocale(const QLocale& locale)
+{
+    return QLocale::languageToString(locale.language()) + " (" + QLocale::countryToString(locale.country()) + ")";
```
Good point. Adjusted now in the new version (and switched from `QLocale::countryToString` deprecated in Qt 6.6 to `QLocale::territoryToString`).

michaelweghorn commented:
> Additionally, I think `qt6-speech-dev` needs to be added the list of packages to install in `.github/workflows/ci.yml` for the CI pipeline to pass.

Thanks. I've adjusted that now. From what I can see, CI isn't triggered automatically on push, but we'll see once it runs...

michaelweghorn commented:
> I've pushed a new version that includes another rebase onto current master including conflict resolution and tries to address the review comments. I've marked most of them as resolved and added questions for 2 for clarification.

PS: My comments weren't actually included the first time I typed them and clicked "Resolve conversation", but I replied with individual comments now to each of them now. -- Sorry for the email noise... (I've only now read https://stackoverflow.com/questions/76388074/github-how-can-i-reply-to-several-comments-in-one-go describing how to do this properly in Github...)

michaelweghorn commented on src/frontend/displaywindow/cdisplaywindow.cpp:
```diff
@@ -551,6 +569,11 @@ QMenu * CDisplayWindow::newDisplayWidgetPopupMenu() {
     popupMenu->addMenu(m_actions.printMenu);
+
+#ifdef BUILD_TEXT_TO_SPEECH
+    popupMenu->addAction((m_actions.speakSelectedText));
```
Good catch. That was indeed missing and is fixed now. (I might have missed this during the last rebase, when things were moved around quite a bit.)

jaakristioja commented on cmake/BTApplication.cmake:
```diff
@@ -14,6 +14,12 @@ FIND_PACKAGE(Qt6 6.7 REQUIRED COMPONENTS
+IF(BUILD_TEXT_TO_SPEECH)
+    FIND_PACKAGE(Qt6 REQUIRED COMPONENTS
+        TextToSpeech
+    )
```
It might be better to fit this `FIND_PACKAGE()` call one line, because it would be less than 80 characters and it is unlikely that new Qt components will be added here any time soon.

michaelweghorn replied:
Done

jaakristioja commented on src/frontend/bibletime.cpp:
```diff
@@ -332,3 +335,39 @@ void BibleTime::openSearchDialog(BtConstModuleList modules,
+    std::unique_ptr<QTextToSpeech> tts;
+
+    // restore settings from config
+    QString const configuredEngine = btConfig().value<QString>(QStringLiteral("GUI/ttsEngine"));
```
As I [explained](https://github.com/bibletime/bibletime/pull/356#pullrequestreview-2660364711) earlier, I think it were better to use `auto const` here and similarly in other places like this. I think it is relatively safe to use `auto` because it is unlikely that the type of expression will change in ways which could cause silent run-time failures, especially the return value of `BtConfig::value<QString>()`. But if you have a different strong opinion on the use of `auto`, I'd be interested in learning about your perspective. :)

For other places I'd like to see `auto` being used, I'll just comment briefly by stating the expected declaration, e.g. `auto &` or `auto const` or similar, so you can more easily spot those places.

michaelweghorn replied:
Thanks for the explicit suggestions, which I've implemented in the newly pushed version.

> But if you have a different strong opinion on the use of auto, I'd be interested in learning about your perspective. :)

While I personally have some preference for explicitly specifying the type (except for types like iterators - `std::vector<SomeClass>::iterator` etc.) as I personally find code a bit more readable/obvious then (type is clear at first glance, while it might need a second/glance thought when it's `auto`), I don't have a strong opinion here on this and code style in general and try to adhere to whatever the code style of the particular project I'm contributing to is using (and am happy to change it when I missed anything at first).

jaakristioja commented on src/frontend/bibletime.cpp:
```diff
@@ -332,3 +335,39 @@ void BibleTime::openSearchDialog(BtConstModuleList modules,
+        tts = std::make_unique<QTextToSpeech>();
+    }
+
+    QLocale const configuredLocale = btConfig().value<QLocale>(QStringLiteral("GUI/ttsLocale"));
```
`auto const`

michaelweghorn replied:
Done

jaakristioja commented on src/frontend/bibletime.cpp:
```diff
@@ -332,3 +335,39 @@ void BibleTime::openSearchDialog(BtConstModuleList modules,
+    if (tts->availableLocales().contains(configuredLocale))
+        tts->setLocale(configuredLocale);
+
+    QString const configuredVoice = btConfig().value<QString>(QStringLiteral("GUI/ttsVoice"));
```
`auto const`

michaelweghorn replied:
Done

jaakristioja commented on src/frontend/bibletime.cpp:
```diff
@@ -332,3 +335,39 @@ void BibleTime::openSearchDialog(BtConstModuleList modules,
+
+    QString const configuredVoice = btConfig().value<QString>(QStringLiteral("GUI/ttsVoice"));
+    QList<QVoice> const voices = tts->availableVoices();
+    for (auto const & voice : voices) {
```
It would be more brief to use `for (auto const & voice : tts->availableVoices()) {` as one would not need to spell out the type (or `auto`).

michaelweghorn replied:
Done

jaakristioja commented on src/frontend/settingsdialogs/bttexttospeechsettings.cpp:
```diff
@@ -0,0 +1,148 @@
+             QLocale::territoryToString(locale.territory()));
+}
+
+}
```
I suggest adding a `// anonymous namespace` comment to the closing brace.

michaelweghorn replied:
Done

jaakristioja commented on src/frontend/settingsdialogs/bttexttospeechsettings.cpp:
```diff
@@ -0,0 +1,148 @@
+        : BtConfigDialog::Page(CResMgr::settings::audio::icon(), parent) {
+
+    QVBoxLayout * mainLayout = new QVBoxLayout(this);
+    QFormLayout * formLayout = new QFormLayout();
```
`auto * const` for these two.

michaelweghorn replied:
Done

jaakristioja commented on src/frontend/settingsdialogs/bttexttospeechsettings.cpp:
```diff
@@ -0,0 +1,148 @@
+    QVBoxLayout * mainLayout = new QVBoxLayout(this);
+    QFormLayout * formLayout = new QFormLayout();
+
+    QStringList const availableEngines = QTextToSpeech::availableEngines();
```
`auto const`

michaelweghorn replied:
Done

jaakristioja commented on src/frontend/settingsdialogs/bttexttospeechsettings.cpp:
```diff
@@ -0,0 +1,148 @@
+void BtTextToSpeechSettingsPage::ttsLocaleChanged() {
+    m_ttsVoiceComboBox->clear();
+
+    QVariant const userData = m_ttsLocaleComboBox->currentData();
```
`auto const`

michaelweghorn replied:
Done

jaakristioja commented on src/frontend/bibletime_slots.cpp:
```diff
@@ -63,6 +63,12 @@ void BibleTime::slotSettingsOptions() {
                    refreshProfileMenus();
                    m_infoDisplay->updateColors();
+
+#ifdef BUILD_TEXT_TO_SPEECH
```
Please indent the preprocessor directives here and in other code blocks as well.

michaelweghorn replied:
Is the indentation correct in the new version?

jaakristioja replied:
Yes, thanks! 😃

jaakristioja commented:
I'm very happy with the changes you've made! And I was finally able to test this on my computer after properly configuring speech-dispatcher. It speaks. :-)

I found two issues. The first issue is a settings dialog page layout problem with the combo box sometimes being too narrow as shown on this screenshot:

![image](https://github.com/user-attachments/assets/5392ad3e-602f-4ff9-9b9a-573a42124721)

Secondly, what was very confusing for me at first is was that the "Speak selected text" popup menu item didn't seem to work. Only later did it occur to me that I had no text selected, duh! I think the item should be grayed out and disabled if there is no text selection (or if the selection doesn't seem to contain anything "speakable").

Mostly for discussion: have you also considered adding a "Speak this" option if the popup menu is opened on a verse?

michaelweghorn commented:
Thanks for looking into this and the additional feedback!
I plan to submit a new version with the comments addressed, but have more questions/thoughts on 2 particular aspects you mention:

> Secondly, what was very confusing for me at first is was that the "Speak selected text" popup menu item didn't seem to work. Only later did it occur to me that I had no text selected, duh! I think the item should be grayed out and disabled if there is no text selection (or if the selection doesn't seem to contain anything "speakable").

I agree disabling the menu entry when there's no selection makes sense.
The logic for enabling/disabling the menu entry is currently the same as for the "Copy" -> "Copy" menu entry that copies the currently selected text to the clipboard, for which I can basically reproduce an equivalent behavior: It is enabled when there is no text selection, and then copies an empty string to the clipboard.

I've submitted https://github.com/bibletime/bibletime/pull/505 to adjust behavior for that menu entry as I think disabling that one also makes sense when there's no selection.
If you agree that this makes sense, I would do a rebase onto (then) current master (and adjust as needed for the new "Speak selected entry" as well) once the discussion there is finished.

> Mostly for discussion: have you also considered adding a "Speak this" option if the popup menu is opened on a verse?

Extending this to allow announcing things other than the current selection sounds good to me and I can look into it. I tend to think it might make sense to keep this somewhat in sync with the "Copy" options for consistency. What do you think? (Would it e.g. make sense to add both, a "Copy this verse" and "Speak this verse" option to copy/read the verse that on which the right-click happened, either as part of this pull request, or potentially as a separate one afterwards to extend the solution further?)

michaelweghorn commented on src/frontend/display/btmodelviewreaddisplay.cpp:
```diff
@@ -248,6 +251,20 @@ void BtModelViewReadDisplay::print(TextPart const type,
+        return;
+
+    BibleTime::instance()->speakText(qml->getSelectedText());
+}
```
Done

michaelweghorn commented:
> The first issue is a settings dialog page layout problem with the combo box sometimes being too narrow as shown on this screenshot:
> [...]

This should now be fixed by the 2 calls to `QComboBox::setSizeAdjustPolicy(QComboBox::AdjustToContents);` added in the new version of this PR. (It at least worked fine now in my tests, but I'd appreciate feedback whether that's true for you, too.)

jaakristioja commented:
> > Secondly, what was very confusing for me at first is was that the "Speak selected text" popup menu item didn't seem to work. Only later did it occur to me that I had no text selected, duh! I think the item should be grayed out and disabled if there is no text selection (or if the selection doesn't seem to contain anything "speakable").
> 
> I agree disabling the menu entry when there's no selection makes sense. The logic for enabling/disabling the menu entry is currently the same as for the "Copy" -> "Copy" menu entry that copies the currently selected text to the clipboard, for which I can basically reproduce an equivalent behavior: It is enabled when there is no text selection, and then copies an empty string to the clipboard.
> 
> I've submitted #505 to adjust behavior for that menu entry as I think disabling that one also makes sense when there's no selection. If you agree that this makes sense, I would do a rebase onto (then) current master (and adjust as needed for the new "Speak selected entry" as well) once the discussion there is finished.

Yes, this sounds reasonable.

> > Mostly for discussion: have you also considered adding a "Speak this" option if the popup menu is opened on a verse?
> 
> Extending this to allow announcing things other than the current selection sounds good to me and I can look into it. I tend to think it might make sense to keep this somewhat in sync with the "Copy" options for consistency. What do you think? (Would it e.g. make sense to add both, a "Copy this verse" and "Speak this verse" option to copy/read the verse that on which the right-click happened, either as part of this pull request, or potentially as a separate one afterwards to extend the solution further?)

In this light, the whole popup menu might need a redesign. Okay, let's keep the "Speak this verse" feature for another time. I would not like to make this PR too complicated.

jaakristioja commented:
> > The first issue is a settings dialog page layout problem with the combo box sometimes being too narrow as shown on this screenshot:
> > [...]
> 
> This should now be fixed by the 2 calls to `QComboBox::setSizeAdjustPolicy(QComboBox::AdjustToContents);` added in the new version of this PR. (It at least worked fine now in my tests, but I'd appreciate feedback whether that's true for you, too.)

Yes, this seems to work now, thanks! :)

michaelweghorn commented:
> In this light, the whole popup menu might need a redesign. Okay, let's keep the "Speak this verse" feature for another time. I would not like to make this PR too complicated.

That sounds good to me, thanks!

michaelweghorn commented:
> On [pics/icons/audio.svg](https://github.com/bibletime/bibletime/pull/356#discussion_r2030001134):
> 
> I don't think it makes sense for me to block this PR simply because of the current mess.
> 
> I guess public domain is okay, but [actual license on the website](https://publicicons.org/license) is a modified version of the [Unlicence](https://unlicense.org/) licence, and the additions are suspicious. The repository providing the actual files instead seems to be licensed under [CC0 1.0 Universal](https://github.com/davidmerfield/Public-Icons/blob/master/LICENSE). This also is suspicious. If you would provide your own it could be under the same GPL-2.0-only as the rest of the BibleTime code.

Thanks for the investigation! I didn't notice any of that, but that indeed looks odd.

I'm afraid creating a proper icon from scratch is beyond my design skills.

I have found this icon on Wikimedia as an alternative, which as far as I can see is actually in the Public Domain:
https://commons.wikimedia.org/wiki/File:Wikisource_speaker_icon.svg

Could that possibly be an option? If simple changes (like changing colors,...) to the existing icon that don't require any particular design skills are desirable, I'm happy to do those.

michaelweghorn commented:
> I'm fine with this public domain icon for the time being, but please make note of the license (e.g. in the commit message if you can't find a better place). :)

I've replaced the icon (and dropped some Inkscape metadata from the original file) now and updated the commit message accordingly.

michaelweghorn commented:
@jaakristioja Any thoughts on the current version of this PR? (I've just done another rebase onto current master + corresponding conflict resolution.)

jaakristioja reviewed (approved):
@michaelweghorn Apologies for the delayed review. You’ve done an outstanding job delivering this feature. I'm truly impressed and very pleased with the work. I see no reason to hold back the merge any longer, so I'm approving the PR. Thanks you so much for your excellent contribution. It's a pleasure to collaborate with you. God bless! :)
